### PR TITLE
Users/ersciple/m108legacyfind

### PR DESCRIPTION
**Pull request by ericsciple**

### CODE REVIEW

**Comment by ericsciple** [comment]
there's a few more unit tests i'm planning to add...

File: node/task.ts [hunk 1]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
 }
 
 /**
+ * Prefer tl.find() and tl.match() instead. This function is for backward compatibility
+ * when porting tasks from PowerShell3 to Node.
+ */
+export function legacyFindFiles(
+        directory: string,
```
Comment (TingluoHuang) [inline node/task.ts]: rootfolder

File: node/task.ts [hunk 2]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
...
+        pattern: string,
+        includeFiles?: boolean,
+        includeDirectories?: boolean,
+        includeHidden?: boolean): string[] {
```
Comment (TingluoHuang) [inline node/task.ts]: no need for includehidden

File: node/task.ts [hunk 3]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
...
+    if (!includeFiles && !includeDirectories) {
+        includeFiles = true;
+    }
+
+    // organize the patterns into include patterns and exclude patterns
+    let includePatterns: string[] = [];
+    let excludePatterns: RegExp[] = [];
+    pattern = pattern.replace(/;;/g, '\0');
+    for (let pat of pattern.split(';')) {
+        if (!pat) {
```
Comment (TingluoHuang) [inline node/task.ts]: what if the pattern is  ";"

File: node/task.ts [hunk 4]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
...
+        else if (pat.startsWith('-:')) {
+            pat = pat.substring(2);
+            isIncludePattern = false;
+        }
+        else {
+            isIncludePattern = true;
+        }
+
+        // validate pattern does not end with a slash
+        if (pat.endsWith('\\' || pat.endsWith('/'))) {
```
Comment (TingluoHuang) [inline node/task.ts]: \\ is only for windows

File: node/task.ts [hunk 5]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
...
+}
+
+function legacyFindFiles_convertPatternToRegExp(pattern: string): RegExp {
+    pattern = (process.platform == 'win32' ? pattern.replace(/\\/g, '/') : pattern) // normalize separator on Windows
+        .replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&') // regex escape - from http://stackoverflow.com/questions/3561493/is-there-a-regexp-escape-function-in-javascript
+        .replace(/\\\/\\\*\\\*\\\//g, '((\/.+/)|(\/))') // replace directory globstar
+        .replace(/\\\*\\\*/g, '.*') // replace remaining globstars with a wildcard that can span directory separators
+        .replace(/\\\*/g, '[^\/]*') // replace asterisks with a wildcard that cannot span directory separators
+        .replace(/\\\?/g, '[^\/]') // replace single character wildcards
+    pattern = `^${pattern}$`;
```
Comment (TingluoHuang) [inline node/task.ts]: add some examples.

File: node/task.ts [hunk 6]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
...
+            .filter((item: string) => {
+                if (includeFiles && includeDirectories) {
+                    return true;
+                }
+
+                let isDir = fs.statSync(item).isDirectory();
+                return (includeFiles && !isDir) || (includeDirectories && isDir);
+            })
+            .forEach((item: string) => {
+                let normalizedPath = item.replace(/\\/g, '/'); // normalize separators
```
Comment (TingluoHuang) [inline node/task.ts]: windows vs linux. / vs \\

File: node/task.ts [hunk 7]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
...
+                if (isMatch) {
+                    allFiles[item] = item;
+                }
+            });
+    }
+
+    return Object.keys(allFiles).sort();
+}
+
+function legacyFindFiles_getDirectoryName(p: string) {
```
Comment (TingluoHuang) [inline node/task.ts]: https://github.com/dotnet/corefx/blob/7bfa0c77da5a14bdef41e7530fee41d5fb3fd912/src/System.Runtime.Extensions/src/System/IO/Path.cs

File: node/task.ts [hunk 8]
```diff
@@ -1093,6 +1092,278 @@ class FindItem {
 }
 
 /**
+ * Prefer tl.find() and tl.match() instead. This function is for backward compatibility
+ * when porting tasks from PowerShell3 to Node.
```
Comment (ericsciple) [inline node/task.ts]: yes, that's true. i'll update

**Pull request merged**